Feature/cells argument

Hoohm opened the pull request.

lldelisle commented:
I was trying to make the code of this branch work and I observed an unexpected behaviour with your choice of 'join_asof' to correct cell barcodes:
```python
In [1]: import polars as pl

In [2]: barcodes_df = pl.DataFrame({'barcode': ["AACATATTCTTTACTG", "TAAAGGGAAGTCAAGC", "TAAATATTCTTTACTG", "TACATATTCTTTACTG", "
   ...: TAGAGCGAAGTCAAGC", "TAGAGGGAAGTCAAGC"], 'count': [1, 1, 1, 98, 1, 98]})

In [3]: barcode_subset_df = pl.DataFrame({'whitelist': ['TACATATTCTTTACTG', 'TAGAGGGAAGTCAAGC']})

In [4]: barcode_subset_df = barcode_subset_df.with_columns(
   ...:             reference=pl.col("whitelist"))

In [5]: barcode_subset_df
Out[5]: 
shape: (2, 2)
┌──────────────────┬──────────────────┐
│ whitelist        ┆ reference        │
│ ---              ┆ ---              │
│ str              ┆ str              │
╞══════════════════╪══════════════════╡
│ TACATATTCTTTACTG ┆ TACATATTCTTTACTG │
│ TAGAGGGAAGTCAAGC ┆ TAGAGGGAAGTCAAGC │
└──────────────────┴──────────────────┘

In [6]: BARCODE_COLUMN = 'barcode'

In [7]: WHITELIST_COLUMN = 'whitelist'

In [8]: temp1 = barcodes_df.sort(BARCODE_COLUMN).join_asof(
   ...:             barcode_subset_df.sort(WHITELIST_COLUMN),
   ...:             left_on=BARCODE_COLUMN,
   ...:             right_on=WHITELIST_COLUMN,
   ...:         )

In [9]: temp1
Out[9]: 
shape: (6, 4)
┌──────────────────┬───────┬──────────────────┬──────────────────┐
│ barcode          ┆ count ┆ whitelist        ┆ reference        │
│ ---              ┆ ---   ┆ ---              ┆ ---              │
│ str              ┆ i64   ┆ str              ┆ str              │
╞══════════════════╪═══════╪══════════════════╪══════════════════╡
│ AACATATTCTTTACTG ┆ 1     ┆ null             ┆ null             │
│ TAAAGGGAAGTCAAGC ┆ 1     ┆ null             ┆ null             │
│ TAAATATTCTTTACTG ┆ 1     ┆ null             ┆ null             │
│ TACATATTCTTTACTG ┆ 98    ┆ TACATATTCTTTACTG ┆ TACATATTCTTTACTG │
│ TAGAGCGAAGTCAAGC ┆ 1     ┆ TACATATTCTTTACTG ┆ TACATATTCTTTACTG │
│ TAGAGGGAAGTCAAGC ┆ 98    ┆ TAGAGGGAAGTCAAGC ┆ TAGAGGGAAGTCAAGC │
└──────────────────┴───────┴──────────────────┴──────────────────┘
```
What I would expect:
```python
┌──────────────────┬───────┬──────────────────┬──────────────────┐
│ barcode          ┆ count ┆ whitelist        ┆ reference        │
│ ---              ┆ ---   ┆ ---              ┆ ---              │
│ str              ┆ i64   ┆ str              ┆ str              │
╞══════════════════╪═══════╪══════════════════╪══════════════════╡
│ AACATATTCTTTACTG ┆ 1     ┆ TACATATTCTTTACTG ┆ TACATATTCTTTACTG │
│ TAAAGGGAAGTCAAGC ┆ 1     ┆ TAGAGGGAAGTCAAGC ┆ TAGAGGGAAGTCAAGC │
│ TAAATATTCTTTACTG ┆ 1     ┆ TACATATTCTTTACTG ┆ TACATATTCTTTACTG │
│ TACATATTCTTTACTG ┆ 98    ┆ TACATATTCTTTACTG ┆ TACATATTCTTTACTG │
│ TAGAGCGAAGTCAAGC ┆ 1     ┆ TAGAGGGAAGTCAAGC ┆ TAGAGGGAAGTCAAGC │
│ TAGAGGGAAGTCAAGC ┆ 98    ┆ TAGAGGGAAGTCAAGC ┆ TAGAGGGAAGTCAAGC │
└──────────────────┴───────┴──────────────────┴──────────────────┘
```

Hoohm commented:
Good catch! Should have started with writing tests!

I can go back to a simpler but slightly slower implementation or find a better way using the asof join.

Hoohm commented:
@lldelisle I rewrote `correct_barcodes_pl` and I added some tests including the ones you ran. I think this time it works. Basically run the asof join twice with both strategies and only keep the ones that have a close enough hamming distance.

Again, thank you for catching this so early!

Hoohm commented:
I've turned off UMI correction for now just to see runs go through completely. The path without cell reference and whitelist should be working. I've still to write tests for the MTX outputs.
